gnuradio.pkgs: init + 3.8 -> 3.9

doronbehar opened the pull request:
- Write a `mkDerivation` and `mkDerivationWith` function for gnuradio,
  like qt5.
- qradiolink, gqrx: Use gnuradio's mkDerivation.
- Use gnuradio's `mkDerivation` to define all gnuradio.pkgs.
- Move all gnuradio packages expressions to `pkgs/development/gnuradio-modules/`.
- Make the gnuradio wrapped and unwrapped derivations compose like python.
- Add more paths to gnuradio's wrapper - add the extra packages as
python modules, and add their executables with proper env vars
wrapping.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Close https://github.com/NixOS/nixpkgs/issues/82263 .

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

doronbehar committed:
gnuradio.pkgs: init

- Write a `mkDerivation` and `mkDerivationWith` function for gnuradio,
  like qt5.
- qradiolink, gqrx: Use gnuradio's callPackage and mkDerivation.
- Use gnuradio.callPackage to define all gnuradio.pkgs.
- Move all gnuradio packages expressions to pkgs/development/gnuradio-modules/ -
  modeled after Python's.
- Add more paths to gnuradio's wrapper - add the extra packages as
  python modules, and add their executables with proper env vars
  wrapping.

Co-authored-by: Frederik Rietdijk <fridh@fridh.nl>

doronbehar committed:
qradiolink: 0.5.0 -> 0.8.5-2

Use gnuradio.pkgs.mkDerivation (gnuradio 3.8) and updated deps.

markuskowa commented:
building `gnuradio-with-packages` fails with `Alias gnuradio-with-packages is still in all-packages.nix`

doronbehar commented:
> building `gnuradio-with-packages` fails with `Alias gnuradio-with-packages is still in all-packages.nix`

Fixed now :), and merge conflict fixed as well.

doronbehar commented on pkgs/applications/radio/gnuradio/pkgs/default.nix:
```diff
@@ -0,0 +1,39 @@
+  };
+in
+  # lib.genAttrs (builtins.attrNames srcs)
+  # (name: callPackage ./${name}.nix {
```
I wonder whether it's possible to do this..

markuskowa commented on pkgs/top-level/all-packages.nix:
```diff
@@ -21250,54 +21244,59 @@ in
-  };
 
   grandorgue = callPackage ../applications/audio/grandorgue { };
+<<<<<<< HEAD
```
This seems to be a leftover conflict marker.

doronbehar replied:
Right, missed that, just fixed it.

markuskowa commented on pkgs/applications/radio/gnuradio/pkgs/srcs.nix:
```diff
@@ -0,0 +1,112 @@
+}:
+
+let
+  knownSrcs = {
```
Using the same code for the derivation but exchanging only the sources between 3.7 and 3.8 relies on the assumption that the newer version does not require any modifications to derivation. How likely is it that this works?

doronbehar replied:
Good question. If you'll look at the modified expressions of the packages, you'll see there's isn't much there at all (besides osmosdr - the most complicated one). The reason they got so simplified is because they are built with `gnuradio.callPackage`'s `mkDerivation` which wraps `stdenv.mkDerivation` with `nativeBuildInputs` and `buildInputs` that are used in the unwrapped gnuradio's derivation - and that usually sums up all that is needed for these packages.

If modifications are needed for a gnuradio package based on the version of gnuradio itself, it's possible to import `gnuradio` in the expression of the package and add conditionals based on `gnuradio.versionAttrs`.

I checked (not with `nixpkgs-review`, but manually) that all attributes that hydra will build, build. For instance the attribute `gnuradioMinimal.pkgs.rds` doesn't build because rds 3.8 needs it's gnuradio (3.8) to be built with swig and python support which `gnuradioMinimal` is built without.

markuskowa replied:
OK, let's give it try.  The osmosdr derivation does not look that bad. 
The Python ecosystem has the same problems to handle. There one can disable a library for a specific version: `disabled = !isPy3k;` Can we do something similar?

doronbehar replied:
Our situation is a bit different. We want to have gnuradio plugins compiled for different gnuradio versions, but all of the upstream authors of these plugins had to release new versions that should work with gr-3.8. They didn't release these versions in a manner compatible with both 3.8 3.7 at the same time. Look at the end of `./pkgs/srcs.nix` to see how we "disable" certain plugins for certain gnuradio packages.

markuskowa replied:
OK, I see.

markuskowa commented on pkgs/applications/radio/qradiolink/default.nix:
```diff
@@ -1,51 +1,64 @@
+mkDerivation rec {
   pname = "qradiolink";
-  inherit version;
+  version = "0.8.5-rc4";
```
Does the new version even build with 0.8.5? Otherwise we could just wait with the update until we have a working gr-3.8.

doronbehar replied:
> Does the new version even build with 0.8.5?

The new version of qradiolink? It builds and it's functional if built with gnuradio 3.8. I just wanted to do most of the update myself in this PR because it needed a lot of refactoring and it illustrates the usage of `gnuradio.callPackage`. I won't mind waiting for a non-rc release.

markuskowa replied:
Let's wait with the upgrade of qradiolink at least until we have a working gr-3.8 (maybe with the next PR then?). Otherwise it breaks qradiolink.

doronbehar replied:
I don't understand why you say it breaks something. qradiolink 0.8.5-rc5 is built with gnuradio 3.8 and it works.

doronbehar replied:
Gnuradio 3.8 also launches and seems functional..

doronbehar committed:
gnss-sdr: Refactor to use gnuradio.pkgs.mkDerivation

Use the same dependencies that gnuradio uses, thanks to gnuradio's
`mkDerivation`. Add `gnuradio.pkgs.osmosdr` as dependency.

doronbehar committed:
gqrx: Refactor to use gnuradioMinimal.pkgs.mkDerivation

- Use gnuradio's `mkDerivation` which includes most of the deps needed.
- Always enable pulseaudio support as that's part of gnuradio's deps
anyway.
- Use gnuradioMinimal.pkgs.osmosdr - not from the alias gr-osmosdr.

doronbehar commented:
Fixed the eval error by pushing two more commits that refactor gqrx and gnss-sdr to use `gnuradio.callPackage` and gnuradio's `mkDerivation`.

doronbehar committed:
inspectrum: Refactor to use gnuradioMinimal.pkgs.mkDerivation

Use the same deps of gnuradioMinimal - thanks to gnuradio's
mkDerivation.

doronbehar committed:
gnuradio: Remove qt5 and gtk from passthru if not used

Attributes such as gnuradioMinimal built without gui support should not
have the `qt` and `gtk` attributes in it's passthru.

doronbehar commented:
Refactored inspectrum as well. Demonstrating that most of the end user applications that need gnuradio can be built with gnuradio 3.8 that itself is built without gui support.

markuskowa commented on pkgs/applications/radio/gnuradio/pkgs/srcs.nix:
```diff
@@ -0,0 +1,113 @@
+{ lib
+, ver
+# Needed for fetching sources
+, pkgs
```
`pkgs` is only needed for the fetchers? Why not just handing over the fetchers?

doronbehar replied:
Yes. Well I could have put each fetcher used in the arguments, but then I would have had to put each fetcher in the arguments of `wrapper.nix` and also inherit each fetcher in the call when I `import ./pkgs/srcs.nix` (in `wrapper.nix`). I too don't like it that much and there's also only 1 fetcher used currently. My intention was that a new definition of a gnuradio package with a source from somewhere else then GitHub will require changing the arguments passed between the different files.

Perhaps it would be better indeed to put the sources inside the expressions themselves, where each fetcher is available because the expressions are evaluated with `callPackage`. What I wanted to avoid was having duplicate code in each gnuradio package that will define a package has only a certain source for 3.7/3.8. I will think about doing it as best as possible.

markuskowa replied:
Would it work to call `./pkgs/srcs.nix` with `callPackages` instead of `import`?
Otherwise I would go with your suggestion no 2, to put the sources inside the package derivations themselves. That is more compact and more readable?

doronbehar replied:
I agree that it's more compact and more readable. I pushed 6c20dd4f91ca76864caa647d19dcb13ba63a5293 now (which will be squashed eventually) that removes `srcs.nix` and there is not much duplicate code going on. `mkDerivation.nix` is where the shared logic of "Don't build this derivation for an incompatible gnuradio" is going on.

FRidh commented on pkgs/applications/radio/qradiolink/default.nix:
```diff
@@ -1,51 +1,64 @@
+, libconfig
+# Needs a gnuradio built with qt gui support
+, gnuradio
+, gnuradioPackages
```
I think its best to avoid passing in both the main package and its package set, to avoid risking incompatible versions

FRidh commented on pkgs/top-level/all-packages.nix:
```diff
@@ -21360,14 +21331,7 @@ in
-      pythonSupport = false;
-    };
-  };
+  gqrx = gnuradioMinimal.callPackage ../applications/radio/gqrx { };
```
I think we should avoid using the `callPackage` from another scope, and instead pass in that scope.

doronbehar commented:
Thanks for the review @FRidh. I'd really like to satisfy your comments, but I don't know how :/ and the issues you present are issues present in other ecosystems so it all leads to https://github.com/NixOS/rfcs/pull/83 ... I wonder whether it'd be too versatile to exactly imitate what's going on in `python-packages.nix`, and for sure 1 thing that I would like to at least know how to imitate is this:

How to create an attribute set like this:

```nix
  pkg1 = callPackage ./pkgs1.nix { };
  pkg2 = callPackage ./pkgs2.nix { };
  # ...
```

Where the `callPackage` of each expression is aware of the other attributes defined like it. Meaning that if `pkg2` depend on `pkg1` you _**don't**_ have to:

```nix
  pkg1 = callPackage ./pkgs1.nix { };
  pkg2 = callPackage ./pkgs2.nix {
    inherit pkg1;
  };
  # ...
```

(Inheritance happens automatically)

I also discussed this with @KarlJoad on https://github.com/NixOS/nixpkgs/commit/4625b75d5a0141107f245f5af9784fecf1c579f5 regarding https://github.com/NixOS/nixpkgs/issues/65398 (@KarlJoad - invited for discussion and for us to perhaps gain some mileage through this PR as well :)).

KarlJoad commented:
> Thanks for the review @FRidh. I'd really like to satisfy your comments, but I don't know how :/ and the issues you present are issues present in other ecosystems so it all leads to [NixOS/rfcs#83](https://github.com/NixOS/rfcs/pull/83) ... I wonder whether it'd be too versatile to exactly imitate what's going on in `python-packages.nix`, and for sure 1 thing that I would like to at least know how to imitate is this:
> 
> How to create an attribute set like this:
> 
> ```nix
>   pkg1 = callPackage ./pkgs1.nix { };
>   pkg2 = callPackage ./pkgs2.nix { };
>   # ...
> ```
> 
> Where the `callPackage` of each expression is aware of the other attributes defined like it. Meaning that if `pkg2` depend on `pkg1` you _**don't**_ have to:
> 
> ```nix
>   pkg1 = callPackage ./pkgs1.nix { };
>   pkg2 = callPackage ./pkgs2.nix {
>     inherit pkg1;
>   };
>   # ...
> ```
> 
> (Inheritance happens automatically)
> 
> I also discussed this with @KarlJoad on [4625b75](https://github.com/NixOS/nixpkgs/commit/4625b75d5a0141107f245f5af9784fecf1c579f5) regarding #65398 (@KarlJoad - invited for discussion and for us to perhaps gain some mileage through this PR as well :)).

Personally, I am of the opinion that the shared information (through `callPackage`) should handle the most general case of a single package. If one package depends on another, then that dependency should be made explicit in the call to `callPackage` However, I imagine that can lead to inconsistencies among packages, and understand if the implicit propagation of those dependencies is the better solution.

I agree with @doronbehar on this topic that the infrastructure for an environment built around a core package is not uniform throughout nixpkgs (I was recommended to look at `python-packages` to figure out how to write `octave-packages`, but with how complex python is, `lua-packages` was easier to understand). I am not quite sure if I can give any insights beyond that for this PR.

FRidh commented:
Had this PR starred in my mailbox, but then forgot about it.

> I was recommended to look at python-packages to figure out how to write octave-packages, but with how complex python is, lua-packages was easier to understand

That is fine! `lua-packages` is based on Python as far as I recall, but it doesn't have the legacy that Python has, so that should be OK.

FRidh commented:
In https://github.com/FRidh/nixpkgs/tree/gnuradio I use `makeScope`

doronbehar commented:
Thanks @FRidh ! I picked your changes and rebased the PR, along with other changes to files' and folders' names - to make it even more like Python's.

There's still now the debate of whether end user applications that use `gnuradio`, should be defined in all-packages.nix with `gnuradio.callPackage` or somewhere else? As far as I understand, the problem is that we don't want them to be part of `gnuradio.pkgs` and we don't like that `gnuradio.callPackage` is used in `all-packages.nix` (out of scope) right?

Would it be more sensible to use a `gnuradio-packages.nix` file and in it `callPackage = gnuradio.callPackage`? Then to `inherit` all of the packages there in `all-packages.nix`? I personally don't like it that much because even now sometimes I use `gnuradioMinimal.callPackage` which is different then `gnuradio.callPackage` - so should I make that `gnuradio-packages.nix` evaluated once with `callPackage = gnuradio.callPackage` and once with `callPackage = gnuradioMinimal.callPackage`? Not all packages are working with both and it wouldn't make sense to build them both..

FRidh commented:
> There's still now the debate of whether end user applications that use gnuradio, should be defined in all-packages.nix with gnuradio.callPackage or somewhere else? As far as I understand, the problem is that we don't want them to be part of gnuradio.pkgs and we don't like that gnuradio.callPackage is used in all-packages.nix (out of scope) right?

Do it like Python. Libraries are in `python-packages.nix`, and their functions have parameters for every library and function that is needed to build the library at hand, and applications are in `all-packages.nix`, and you need to pass in the interpreter or package set instead of individual packages. That way, you just use the `callPackage` of `all-packages.nix`.

doronbehar commented:
> Do it like Python. Libraries are in `python-packages.nix`, and their functions have parameters for every library and function that is needed to build the library at hand, and applications are in `all-packages.nix`, and you need to pass in the interpreter or package set instead of individual packages. That way, you just use the `callPackage` of `all-packages.nix`.

Got it. Done!

doronbehar commented:
BTW @markuskowa qradiolink is updated here to a non-rc version now.

doronbehar commented on pkgs/development/gnuradio-modules/mkDerivation.nix:
```diff
@@ -0,0 +1,21 @@
+else
+
+let
+  args_ = {
+    enableParallelBuilding = args.enableParallelBuilding or true;
+    nativeBuildInputs = (args.nativeBuildInputs or []) ++ unwrapped.nativeBuildInputs;
+    buildInputs = (args.buildInputs or []) ++ unwrapped.buildInputs ++ [ unwrapped ];
+  };
+in mkDerivation (args // args_)
```
This is something I don't feel entirely confident with - adding all of those `buildInputs` unconditionally. Is it too verbose? As part of that, `cmake` is added as well, which influences the default build system stdenv will try to build the derivation with. Maybe there should be a `buildGnuradioPackage` and `buildGnuradioApplication` with different behaviors? 

OTH, the situation with most applications and gnuradio plugins as of currently - they need almost all of those deps and some of them don't even list the deps they need in their readme because they expect them to be present if you have Gnuradio installed.

FRidh replied:
indeed, adding all the inputs of gnuradio to all its packages (essentially a limited form of propagation) seems a bit far. Maybe if there are documented packages that are always required, then they can be explicitly added here in `args`, but just taking in all `(native)buildInputs` is I think not OK. I typically prefer to be correct over pragmatic though.

doronbehar commented on pkgs/development/gnuradio-modules/nacl/default.nix:
```diff
@@ -0,0 +1,29 @@
+    rev = "15276bb0fcabf5fe4de4e58df3d579b5be0e9765";
+    sha256 = "018np0qlk61l7mlv3xxx5cj1rax8f1vqrsrch3higsl25yydbv7v";
+  };
+  disabled = ["3.8"];
```
This also a point I'm not sure I chose the right terms and conditionals. In the future there should come another backwards incompatible Gnuradio version 3.9 where this list will need to be extended. These kind of gnuradio plugins will probably never get a new commit and perhaps there should be both a supported `disabled` list and `enabledOnly` variable? to be respected by `gnuradio.pkgs.mkDerivation`? The downside is that it'll increase the complexity there.

FRidh replied:
Why not using boolean logic as is done with Python?

doronbehar replied:
You mean that this will be used inside packages' expressions:

```nix
  disabledIf = isGR3_8 or isGR3_9;
```

? It feels a bit better but I don't see how it is less complex..

FRidh replied:
What is complex about it?

doronbehar replied:
I don't know whether using boolean logic or the current logic is more or less complex. Maybe either options is not complex.. I only expressed a slight dissatisfaction from the current situation, and I don't see a strong benefit in using boolean logic like Python over what's currently used. Maybe it's not that important.

doronbehar replied:
I finally chose using the attribute: `disabledForGRafter = "3.8"` which should not require a change once 3.9 will get in.

doronbehar commented on pkgs/development/gnuradio-modules/osmosdr/default.nix:
```diff
@@ -0,0 +1,50 @@
+}:
+
+let
+  version = {
+    "3.7" = "0.1.5";
+    "3.8" = "0.2.2";
+  }.${gnuradio.versionAttr.major};
+  src = fetchgit {
+    url = "git://git.osmocom.org/gr-osmosdr";
+    rev = "v${version}";
+    sha256 = {
+      "3.7" = "0bf9bnc1c3c4yqqqgmg3nhygj6rcfmyk6pybi27f7461d2cw1drv";
+      "3.8" = "HT6xlN6cJAnvF+s1g2I1uENhBJJizdADlLXeSD0rEqs=";
+    }.${gnuradio.versionAttr.major};
+  };
```
Considering the complexity of `disabled` vs `enabledOnly`, I'm not even sure what would be the fate of these derivations when 3.9 will come..

doronbehar reviewed (commented):
cc @FRidh for an opinion.

eduardosm commented:
I made [a branch based on this one](https://github.com/eduardosm/nixpkgs/tree/gnuradio.pkgs), rebased on master to fix conflicts.

eduardosm commented:
@doronbehar Are you still working on this? Otherwise, I can continue it.

doronbehar commented:
Thanks for offering help @eduardosm . I rebased the current PR with the merge conflicts fixed based on your conflict resolution. There's still some unresolved reviews which perhaps should not block this PR and we could handle these issues in the future. In anycase I'd like to get an approval from @FRidh or someone else more experienced then me with packages sets.

wirew0rm commented:
I also tried to use this branch to build some out of tree modules and one thing I could not figure out was how to handle modules which depend on other modules. I don't know if there are any other modules which need a gnuradio with other modules already present to build. Or maybe it is trivial and I just didn't find the right idiom to use?

doronbehar committed:
volk: init at 2.4.1

doronbehar committed:
gnuradio: 3.8 -> 3.9

Add some "3.9" attributes to srcs in gnuradio packages And update
packages using GR3.8 and that are incompatible yet with GR3.9 to use
GR3.8 explicitly.

eduardosm commented:
gnss-sdr is failing to build due to https://github.com/boostorg/format/issues/73. This can be fixed by building with latest boost, which requires patching gnuradio.

```diff
diff --git a/pkgs/applications/radio/gnuradio/default.nix b/pkgs/applications/radio/gnuradio/default.nix
index b707b361213..fd7299bd5db 100644
--- a/pkgs/applications/radio/gnuradio/default.nix
+++ b/pkgs/applications/radio/gnuradio/default.nix
@@ -29,6 +29,7 @@
 , pango
 , gobject-introspection
 , cairo
+, icu
 , qt5
 , libsForQt5
 # Features available to override, the list of them is in featuresInfo. They
@@ -56,11 +57,13 @@ let
         pkg-config
         orc
       ];
+      # when gr-qtgui is disabled, icu needs to be included, otherwise
+      # building with boost 1.7x fails
       runtime = [
         boost
         log4cpp
         mpir
-      ];
+      ] ++ lib.optionals (!(hasFeature "gr-qtgui" features)) [ icu ];
       pythonNative = with python.pkgs; [
         Mako
         six
@@ -266,6 +269,12 @@ let
       url = "https://github.com/gnuradio/gnuradio/commit/dbc8ad7e7361fddc7b1dbc267c07a776a3f9664b.diff";
       sha256 = "tQcCpcUbJv3yqAX8rSHN/pAuBq4ueEvoVo7sNzZGvf4=";
     })
+    # Needed to use boost 1.7x
+    # https://github.com/gnuradio/gnuradio/issues/3720
+    (fetchpatch {
+      url = "https://github.com/gnuradio/gnuradio/commit/cbcb968358fad56f3646619b258f18b0e6693a07.diff";
+      sha256 = "1ajf4797f869lqv436xw61s29qdbn7f01i0970kfxv3yahd34p9v";
+    })
   ];
 in
 
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index ce1c0412fe3..7257d9709b4 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -22044,6 +22044,9 @@ in
   gnuradio = callPackage ../applications/radio/gnuradio/wrapper.nix {
     unwrapped = callPackage ../applications/radio/gnuradio {
       inherit (darwin.apple_sdk.frameworks) CoreAudio;
+      # boost17x is needed to build gnss-sdr with gnuradio 3.8
+      # https://github.com/boostorg/format/issues/73
+      boost = boost17x;
       python = python3;
     };
   };
```

doronbehar committed:
uhd: 3.15.0.0 -> 4.0.0.0

doronbehar committed:
gnuradio, uhd: Use boost17x

gnss-sdr needs it: https://github.com/boostorg/format/issues/67 .
uhd and gnuradio need to use the same boost version to avoid
incompatibilities issues. icu is needed from some reason with boost17x.

doronbehar commented:
Thanks @eduardosm for noticing and investigating. I verified your findings and I've applied a commit fixing this while also handling `uhd` which also switched to `boost17x` now, and got upgraded. Where did you find the commit https://github.com/gnuradio/gnuradio/commit/cbcb968358fad56f3646619b258f18b0e6693a07 referenced in the issue https://github.com/gnuradio/gnuradio/issues/3720 ? I'd like to add it to the comment near the patch.

eduardosm commented:
The commit is from PR https://github.com/gnuradio/gnuradio/pull/3967. You can find it referenced in https://github.com/gnuradio/gnuradio/issues/3720 just before the last comment.

eduardosm commented:
Result of `nixpkgs-review pr 106599` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>17 packages failed to build:</summary>
  <ul>
    <li>gnss-sdr</li>
    <li>gnuradio3_7</li>
    <li>gnuradio3_7Minimal</li>
    <li>gnuradio3_7Packages.ais</li>
    <li>gnuradio3_7Packages.gnuradio</li>
    <li>gnuradio3_7Packages.gsm</li>
    <li>gnuradio3_7Packages.limesdr</li>
    <li>gnuradio3_7Packages.nacl</li>
    <li>gnuradio3_7Packages.osmosdr</li>
    <li>gnuradio3_7Packages.rds</li>
    <li>gnuradioMinimal</li>
    <li>gnuradioPackages.limesdr</li>
    <li>gnuradioPackages.osmosdr</li>
    <li>gnuradioPackages.rds</li>
    <li>gqrx</li>
    <li>inspectrum</li>
    <li>qradiolink</li>
  </ul>
</details>
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>cubicsdr</li>
    <li>gnuradio</li>
    <li>gnuradioPackages.gnuradio</li>
    <li>python37Packages.soapysdr-with-plugins</li>
    <li>python38Packages.soapysdr-with-plugins</li>
    <li>python39Packages.soapysdr-with-plugins</li>
    <li>rtl_433</li>
    <li>sdrangel</li>
    <li>soapysdr-with-plugins</li>
    <li>soapyuhd</li>
    <li>uhd</li>
    <li>welle-io</li>
  </ul>
</details>

eduardosm commented:
Regarding failing builds:

* gnuradio 3.7 seems to be incompatible with UHD 4.0
* gnuradio 3.8 minimal fails because you forgot the following part of the patch:



> ```diff
> @@ -56,11 +57,13 @@ let
>          pkg-config
>          orc
>        ];
> +      # when gr-qtgui is disabled, icu needs to be included, otherwise
> +      # building with boost 1.7x fails
>        runtime = [
>          boost
>          log4cpp
>          mpir
> -      ];
> +      ] ++ lib.optionals (!(hasFeature "gr-qtgui" features)) [ icu ];
>        pythonNative = with python.pkgs; [
>          Mako
>          six
> ```

eduardosm commented:
Other packages also fail to build, but only after rebasing on master, I'll try to figure it out.

eduardosm commented:
I fixed most (if not all) failures in [my branch](https://github.com/eduardosm/nixpkgs/tree/gnuradio.pkgs) (note that it has been rebased).

doronbehar commented:
I took a slightly different approach towards the `dontWrapQtApps` issue - by not adding by default `qt5.qtbase` to `buildInputs` of GR derivations unconditionally. I also figured out why some packages failed in `master` and not in my branch - https://github.com/NixOS/nixpkgs/pull/113515#issuecomment-787125558 . I rebased everything and ran a `nixpkgs-review` locally before pushing. It should all be ready now for a final approval.

jhenahan commented:
Darwin appears to have been hanging for >8 days. I'd love to see this land. Is there a way to kick ofborg over to try again?

doronbehar commented:
> Darwin appears to have been hanging for >8 days. I'd love to see this land. Is there a way to kick ofborg over to try again?

@SuperSandro2000 for a nixpkgs-review on Darwin :pray: ?

SuperSandro2000 commented on pkgs/applications/radio/gnss-sdr/default.nix:
```diff
@@ -29,27 +30,30 @@ stdenv.mkDerivation rec {
+  nativeBuildInputs = [
+    cmake
+    gnuradio.unwrapped.python
+    gnuradio.unwrapped.python.pkgs.Mako
```
Maybe use with here.

SuperSandro2000 commented on pkgs/applications/radio/gnuradio/default.nix:
```diff
@@ -60,7 +62,10 @@ let
+      ]
+        # when gr-qtgui is disabled, icu needs to be included, otherwise
+        # building with boost 1.7x fails
+        ++ lib.optionals (!(hasFeature "gr-qtgui" features)) [ icu ];
```
```suggestion
        ++ lib.optionals !(hasFeature "gr-qtgui" features) [ icu ];
```

that should also work.

doronbehar replied:
> that should also work.

Doesn't work.

SuperSandro2000 commented on pkgs/applications/radio/gqrx/default.nix:
```diff
@@ -18,10 +26,27 @@ mkDerivation rec {
+    rtl-sdr
+    hackrf
+  ];
+
+  enableParallelBuilding = true;
```
```suggestion
```

set by cmake.

SuperSandro2000 commented on pkgs/applications/radio/uhd/3.5.nix:
```diff
@@ -0,0 +1,158 @@
+, enableExamples ? false
+, enableUtils ? false
+, enableLiberio ? false
+, liberio ? null
+, libusb1 ? null
```
```suggestion
, liberio
, libusb1
```

SuperSandro2000 commented on pkgs/applications/radio/uhd/3.5.nix:
```diff
@@ -0,0 +1,158 @@
+, liberio ? null
+, libusb1 ? null
+, enableDpdk ? false
+, dpdk ? null
```
```suggestion
, dpdk
```

SuperSandro2000 commented on pkgs/applications/radio/uhd/3.5.nix:
```diff
@@ -0,0 +1,158 @@
+{ lib, stdenv
```
```suggestion
{ lib
, stdenv
```

SuperSandro2000 commented on pkgs/development/gnuradio-modules/ais/default.nix:
```diff
@@ -0,0 +1,45 @@
+    description = "Gnuradio block for ais";
+    homepage = "https://github.com/bistromath/gr-ais";
+    license = licenses.gpl3Plus;
+    platforms = platforms.linux ++ platforms.darwin;
```
```suggestion
    platforms = platforms.unix;
```

SuperSandro2000 commented on pkgs/development/gnuradio-modules/rds/default.nix:
```diff
@@ -0,0 +1,56 @@
+    description = "Gnuradio block for radio data system";
+    homepage = "https://github.com/bastibl/gr-rds";
+    license = licenses.gpl2Plus;
+    platforms = platforms.linux ++ platforms.darwin;
```
```suggestion
    platforms = platforms.unix;
```

SuperSandro2000 commented on pkgs/development/gnuradio-modules/osmosdr/default.nix:
```diff
@@ -0,0 +1,82 @@
+  meta = with lib; {
+    description = "Gnuradio block for OsmoSDR and rtl-sdr";
+    homepage = "https://sdr.osmocom.org/trac/wiki/GrOsmoSDR";
+    license = licenses.gpl3Plus;
```
missing platforms.

SuperSandro2000 commented:
> @SuperSandro2000 for a nixpkgs-review on Darwin 🙏 ?

Running it when it is free.

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 106599` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
9 packages marked as broken and skipped:
</summary>
<ul>
<li>gnuradio</li>
<li>gnuradio3_7</li>
<li>gnuradio3_7Minimal</li>
<li>gnuradio3_7Packages.ais</li>
<li>gnuradio3_7Packages.gnuradio</li>
<li>gnuradio3_7Packages.nacl</li>
<li>gnuradio3_7Packages.rds</li>
<li>rtl_433</li>
<li>welle-io</li>
</ul>
</details>
<details>

  <summary>3 packages failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>gnuradioMinimal: log was empty</li>
<li>gnuradioPackages.gnuradio: log was empty</li>
<li>gnuradioPackages.rds: log was empty</li>
  </ul>
</details>
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>uhd: <a href="https://termbin.com/t6lw">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/42e4e2a0195721b080221aedf4758dcc/raw/index.html">pretty log </a></li>
  </ul>
</details>
<details>
<summary>
1 package built:
</summary>
<ul>
<li>uhd3_5</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
gnuradioPackages.gnuradio:
<details>

warning: EvalError
Cannot evaluate attribute ‘gnuradioPackages.gnuradio’ in ‘/nix/store/4npw4zzb4msr75rfbrjb4rc37jqvwnhp-nixpkgs’.
</details>
gnuradioPackages.rds:
<details>

warning: EvalError
Cannot evaluate attribute ‘gnuradioPackages.rds’ in ‘/nix/store/4npw4zzb4msr75rfbrjb4rc37jqvwnhp-nixpkgs’.
</details>
uhd:
<details>

warning: EvalError
Cannot evaluate attribute ‘uhd’ in ‘/nix/store/4npw4zzb4msr75rfbrjb4rc37jqvwnhp-nixpkgs’.
</details>
</details>

eduardosm commented:
UHD failure might be related to https://github.com/EttusResearch/uhd/issues/397.

doronbehar committed:
gnuradio3_{7,8}: Use external volk

Now that volk is packaged, we can use our build and save some closure
space for those that have more then 1 version of gnuradio installed.
Also, simplify a bit how attributes are inherited in the expression.

doronbehar commented:
Now GR 3.9 is here :).

I think there's not much left to do here - It would be great if some users will make sure basic functionality is working with GR3.9 and the gnuradio packages which currently compile only with GR3.8. This can be tested with e.g:

```nix
  gnuradioEnv = gnuradio3_8.override {
    extraPackages = lib.attrVals [
      "osmosdr" "rds" "limesdr"
    ] gnuradio3_8Packages;
  };
```

Checking that GR and all new packages introduced here also work on Darwin would be great, but shouldn't block this from getting merged IMO.

The packages interface has improved since the last reviews and it can be changed in the future if a change to the interface will be required.

FRidh commented:
@doronbehar if you don't get any further feedback despite cc'ing previous contributors, then just go ahead with it.

doronbehar commented:
> @doronbehar if you don't get any further feedback despite cc'ing previous contributors, then just go ahead with it.

Will merge once nixpkgs-review finishes on my machine.

doronbehar commented:
Result of `nixpkgs-review pr 106599` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>gnuradioMinimal</li>
  </ul>
</details>
<details>
  <summary>33 packages built:</summary>
  <ul>
    <li>cubicsdr</li>
    <li>gnss-sdr</li>
    <li>gnuradio</li>
    <li>gnuradio3_7</li>
    <li>gnuradio3_7Minimal</li>
    <li>gnuradio3_7Packages.ais</li>
    <li>gnuradio3_7Packages.gnuradio</li>
    <li>gnuradio3_7Packages.gsm</li>
    <li>gnuradio3_7Packages.limesdr</li>
    <li>gnuradio3_7Packages.nacl</li>
    <li>gnuradio3_7Packages.osmosdr</li>
    <li>gnuradio3_7Packages.rds</li>
    <li>gnuradio3_8</li>
    <li>gnuradio3_8Minimal</li>
    <li>gnuradio3_8Packages.gnuradio</li>
    <li>gnuradio3_8Packages.limesdr</li>
    <li>gnuradio3_8Packages.osmosdr</li>
    <li>gnuradio3_8Packages.rds</li>
    <li>gnuradioPackages.gnuradio</li>
    <li>gqrx</li>
    <li>inspectrum</li>
    <li>pothos</li>
    <li>python38Packages.soapysdr-with-plugins</li>
    <li>python39Packages.soapysdr-with-plugins</li>
    <li>qradiolink</li>
    <li>rtl_433</li>
    <li>sdrangel</li>
    <li>soapysdr-with-plugins</li>
    <li>soapyuhd</li>
    <li>uhd</li>
    <li>uhd3_5</li>
    <li>volk</li>
    <li>welle-io</li>
  </ul>
</details>

doronbehar commented:
Fixed the build of `gnuradioMinimal` but didn't run `nixpkgs-review` again. Merging when ofborg is green.
